### PR TITLE
Reexports styled-jsx JSXStyle in nextjs

### LINKED ISSUE #9325: Non-hoisted styled-jsx dependency breaks styles
# Bug report

## Describe the bug

When `styled-jsx` with a version other than the one bundled with Next.js is manually added as a dependency to a Next.js project, style loading is delayed leading to an unstyled initial rendering flash.

## To Reproduce

Reproduction in: https://github.com/ctavan/next.js-reproduce-styled-jsx-bug/commits/master

Adding `styled-jsx` as a dependency manually produces the buggy behavior: https://github.com/ctavan/next.js-reproduce-styled-jsx-bug/commit/489616fc9156ee9a089eb25f646e4ef10a037e36

## Expected behavior

Adding `styled-jsx` as a dependency manually should either not produce the bug or lead to a build-time error/warning in Next.js (or it should be at least documented that you should not add another version than the bundled one of `styled-jsx` to a Next.js project).

## Screenshots

Here's how the `<head>` looks like, when there is NO additional `styled-jsx` dependency. Note that the `<style>` tags are included _before_ `chunk/0.js` is loaded in a blocking fashion:

working

Here's how the `<head>` looks like, when there is an additional `styled-jsx` dependency (different version than the one bundled with Next.js). Note that the `<style>` tags are included _after_ `chunk/0.js` is loaded in a blocking fashion leading to html content flashing unstyled until `chunk/0.js` has been loaded:

broken

## System information

- OS: macOS
- Version of Next.js: 9.1.2

### LINKED ISSUE #10149: The styled-jsx integration relies on hoisting
# Bug report

## Describe the bug

Next generates files that contain the following import statement:

```ts
import foo from 'styled-jsx/style';
```

Since the output files belong to the project that depends on Next rather than Next itself, the `styled-jsx` dependency needs to be listed in the user project. Based on #9325, I think this isn't what you're looking for.

## Suggested fix

1. Fix `styled-jsx` by:
    - adding a custom import path setting (it's currently [hardcoded](https://github.com/zeit/styled-jsx/blob/a736e3cc8ab5c7ce3db4f01beb0a3f9ee343cc3b/src/_utils.js#L681)).
2. Fix `next` by:
    - adding a new file (let's say `styled-jsx.js`) that contains `module.exports = require('styled-jsx/style')` (it would basically re-export its own dependency)
    - updating the Babel configuration to use `next/styled-jsx` inside the custom import path setting that we will have implemented in 1.

This solution isn't obtrusive, doesn't have breaking changes, and makes Next resilient to problems highlighted in #9325 (and, more importantly to me, makes the default `create-next-app` setup work with Yarn 2 😉).

Note that I likely won't have the resources to make the fix myself right now.

### PR DESCRIPTION
**Pull request by huozhi**
When using pnpm / yarnPnP to install next.js, styled-jsx as dependency is not hoisted in the top level node_modules, it will fail when nodejs is trying to resolve `styled-jsx/style` from project directory. Re-export `styled-jsx/style` in next.js and let swc/babel plugin compile the import path it to `next/dist/shared/lib/styled-jsx`

Resolves #10149
Closes #21320
Closes #9325

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 273f1d8fb661be09ea8aa7b2c048c55c84d0f53f

`yarn testheadless test/integration/app-document-style-fragment/test/index.test.js` 
- Custom Document Fragment Styles > correctly adds styles from fragment styles key

<details>
<summary>Expand output</summary>

  ● Custom Document Fragment Styles › correctly adds styles from fragment styles key

    expect(received).toMatch(expected)

    Expected pattern: /font-size:(.*|)16\.4px/
    Received string:  "html { background: hotpink; }
                    body { margin: 0; color: #000; background: #fff; }
                    .next-error-h1 {
                      border-right: 1px solid rgba(0, 0, 0, .3);
                    }
                    @media (prefers-color-scheme: dark) {
                      body { color: #fff; background: #000; }
                      .next-error-h1 {
                        border-right: 1px solid rgba(255, 255, 255, .3);
                      }
                    }"

      36 |     const styles = $('style').text()
      37 |     expect(styles).toMatch(/background:(.*|)hotpink/)
    > 38 |     expect(styles).toMatch(/font-size:(.*|)16\.4px/)
         |                    ^
      39 |   })
      40 | })
      41 |

      at Object.<anonymous> (integration/app-document-style-fragment/test/index.test.js:38:20)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | huozhi/next.js reexports-styled-jsx | Change |
| - | - | - | - |
| buildDuration | 15.8s | 15.7s | -139ms |
| buildDurationCached | 6.3s | 6.3s | ⚠️ +71ms |
| nodeModulesSize | 463 MB | 463 MB | ⚠️ +423 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | huozhi/next.js reexports-styled-jsx | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.756 | 3.762 | ⚠️ +0.01 |
| / avg req/sec | 665.57 | 664.61 | ⚠️ -0.96 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.396 | 1.407 | ⚠️ +0.01 |
| /error-in-render avg req/sec | 1791.21 | 1777.3 | ⚠️ -13.91 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js reexports-styled-jsx | Change |
| - | - | - | - |
| 925.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42 kB | 42 kB | ✓ |
| main-HASH.js gzip | 28.6 kB | 28.6 kB | ✓ |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 72.3 kB | 72.3 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js reexports-styled-jsx | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js reexports-styled-jsx | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.36 kB | 1.36 kB | ✓ |
| _error-HASH.js gzip | 193 B | 193 B | ✓ |
| amp-HASH.js gzip | 308 B | 308 B | ✓ |
| css-HASH.js gzip | 327 B | 327 B | ✓ |
| dynamic-HASH.js gzip | 3.08 kB | 3.08 kB | ✓ |
| head-HASH.js gzip | 359 B | 359 B | ✓ |
| hooks-HASH.js gzip | 920 B | 920 B | ✓ |
| image-HASH.js gzip | 5.71 kB | 5.71 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 2.64 kB | 2.64 kB | ✓ |
| routerDirect..HASH.js gzip | 320 B | 320 B | ✓ |
| script-HASH.js gzip | 391 B | 391 B | ✓ |
| withRouter-HASH.js gzip | 318 B | 318 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 16.3 kB | 16.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js reexports-styled-jsx | Change |
| - | - | - | - |
| _buildManifest.js gzip | 459 B | 459 B | ✓ |
| Overall change | 459 B | 459 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js reexports-styled-jsx | Change |
| - | - | - | - |
| index.html gzip | 533 B | 533 B | ✓ |
| link.html gzip | 547 B | 547 B | ✓ |
| withRouter.html gzip | 529 B | 529 B | ✓ |
| Overall change | 1.61 kB | 1.61 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | huozhi/next.js reexports-styled-jsx | Change |
| - | - | - | - |
| buildDuration | 17.6s | 17.9s | ⚠️ +235ms |
| buildDurationCached | 6.3s | 6.2s | -81ms |
| nodeModulesSize | 463 MB | 463 MB | ⚠️ +423 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | huozhi/next.js reexports-styled-jsx | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.672 | 3.701 | ⚠️ +0.03 |
| / avg req/sec | 680.89 | 675.58 | ⚠️ -5.31 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.35 | 1.351 | 0 |
| /error-in-render avg req/sec | 1851.52 | 1851.16 | ⚠️ -0.36 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js reexports-styled-jsx | Change |
| - | - | - | - |
| 925.HASH.js gzip | 178 B | 178 B | ✓ |
| framework-HASH.js gzip | 42.2 kB | 42.2 kB | ✓ |
| main-HASH.js gzip | 29 kB | 29 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 72.9 kB | 72.9 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js reexports-styled-jsx | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js reexports-styled-jsx | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.35 kB | 1.35 kB | ✓ |
| _error-HASH.js gzip | 179 B | 179 B | ✓ |
| amp-HASH.js gzip | 312 B | 312 B | ✓ |
| css-HASH.js gzip | 324 B | 324 B | ✓ |
| dynamic-HASH.js gzip | 3.08 kB | 3.08 kB | ✓ |
| head-HASH.js gzip | 357 B | 357 B | ✓ |
| hooks-HASH.js gzip | 921 B | 921 B | ✓ |
| image-HASH.js gzip | 5.76 kB | 5.76 kB | ✓ |
| index-HASH.js gzip | 261 B | 261 B | ✓ |
| link-HASH.js gzip | 2.76 kB | 2.76 kB | ✓ |
| routerDirect..HASH.js gzip | 322 B | 322 B | ✓ |
| script-HASH.js gzip | 392 B | 392 B | ✓ |
| withRouter-HASH.js gzip | 317 B | 317 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 16.5 kB | 16.5 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js reexports-styled-jsx | Change |
| - | - | - | - |
| _buildManifest.js gzip | 458 B | 458 B | ✓ |
| Overall change | 458 B | 458 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | huozhi/next.js reexports-styled-jsx | Change |
| - | - | - | - |
| index.html gzip | 528 B | 528 B | ✓ |
| link.html gzip | 541 B | 541 B | ✓ |
| withRouter.html gzip | 525 B | 525 B | ✓ |
| Overall change | 1.59 kB | 1.59 kB | ✓ |

</details>

</details>
Commit: f63e5c8941568ae8eb0a46f1c847f0d37eebb71f

File: .github/workflows/build_test_deploy.yml [hunk 1]
```diff
@@ -629,6 +629,9 @@ jobs:
       - run: xvfb-run node run-tests.js --timings -g ${{ matrix.group }}/18
         if: ${{needs.build.outputs.docsChange != 'docs only change'}}
 
+      - run: NEXT_TEST_MODE=start PKG_MANAGER=pnpm xvfb-run node run-tests.js test/e2e/styled-jsx/index.test.ts
```
Comment (timneutkens) [inline .github/workflows/build_test_deploy.yml:632]: This doesn't seem ideal, I think the more ideal way to go about this is to run 2 tests with different inputs. I'll update `test/e2e/styled-jsx/index.test.ts` to reflect that.

Comment (timneutkens) [reply .github/workflows/build_test_deploy.yml]: Updated to this approach: https://github.com/vercel/next.js/pull/36585/files#diff-e1b72b64ea6f1f2833924b8ce31283c68e6df60bd064fc4a312a968b82f79839R42

**Review by timneutkens — APPROVED** [review]

**Pull request merged**